code/modules/research/ordnance/tank_compressor.dm

```diff
@@ -0,0 +1,330 @@
+
+/// Stores the record of the gas data for a significant enough tank leak
+/datum/data/compressor_record
+	/// Tank Name
+	var/experiment_source
+	/// Key: Path, Value: Moles
+	var/list/gas_data = list()
+	var/timestamp
+
+/obj/machinery/atmospherics/components/binary/tank_compressor/attackby(obj/item/item, mob/living/user)
+	if (!panel_open)
+		if(istype(item, /obj/item/tank))
+			var/obj/item/tank/tank_item = item
+			if(inserted_tank)
+				if(!eject_tank(user))
+					balloon_alert(user, span_warning("[inserted_tank] is stuck inside."))
+					return ..()
+			if(!user.transferItemToLoc(tank_item, src))
+				balloon_alert(user, span_warning("[tank_item] is stuck to your hand."))
+				return ..()
+			inserted_tank = tank_item
+			last_recorded_pressure = 0
+			RegisterSignal(inserted_tank, COMSIG_PARENT_QDELETING, .proc/unregister_tank)
+			update_appearance()
+			return
+		if(istype(item, /obj/item/computer_hardware/hard_drive/portable))
+			var/obj/item/computer_hardware/hard_drive/portable/attacking_disk = item
+			eject_disk(user)
+			if(user.transferItemToLoc(attacking_disk, src))
+				inserted_disk = attacking_disk
+			else
+				balloon_alert(user, span_warning("[attacking_disk] is stuck to your hand."))
+			return
+	if(!active && !inserted_tank)
+		if(default_deconstruction_screwdriver(user, "[base_icon_state]-open", "[base_icon_state]-open", item))
+			update_appearance()
+			return
+		if(panel_open)
+			if(default_deconstruction_crowbar(item))
+				return
+			if(default_change_direction_wrench(user, item))
+				return
+	return ..()
+
+/obj/machinery/atmospherics/components/binary/tank_compressor/default_change_direction_wrench(mob/user, obj/item/wrench)
+	. = ..()
+	if(.)
+		/// Disconnect our partner.
+		if(nodes[1])
+			nodes[1].disconnect(src)
+			nodes[1] = null
+			if(parents[1])
+				nullify_pipenet(parents[1])
+		if(nodes[2])
+			nodes[2].disconnect(src)
+			nodes[2] = null
+			if(parents[2])
+				nullify_pipenet(parents[2])
+		set_init_directions()
+		/// Connect to a new one.
+		atmos_init()
+		if(nodes[1])
+			nodes[1].atmos_init()
+			nodes[1].add_member(src)
+		if(nodes[2])
+			nodes[2].atmos_init()
+			nodes[2].add_member(src)
+		SSair.add_to_rebuild_queue(src)
+		update_appearance()
+
+/// Glorified volume pump.
+/obj/machinery/atmospherics/components/binary/tank_compressor/process_atmos()
+	if(!active || !airs[2].total_moles() || !transfer_rate || !inserted_tank)
+		return
+
+	var/datum/gas_mixture/tank_air = inserted_tank.return_air()
+	if(airs[2].return_pressure() < 0.01 || tank_air.return_pressure() > TANK_COMPRESSOR_PRESSURE_LIMIT)
+		return
+
+	/// Prevent pumping if tank is taking damage but still below pressure limit. Here to prevent exploiting the buffer system.
+	if((inserted_tank.leaking) && (tank_air.return_pressure() <= TANK_LEAK_PRESSURE))
+		active = FALSE
+		return
+
+	var/datum/gas_mixture/removed = airs[2].remove_ratio(transfer_rate / airs[2].volume)
```

Comment:
A: can you be guarenteed that airs[2] exists?
B: rather then accessing the mix from the airs list each time, could you make a var to hold it? I think it's a bit nicer

Reply:
A: Ill add an if check
B: Yeah sure